src/lib/storage/dictionary_segment.cpp

```diff
@@ -0,0 +1,133 @@
+#include <unordered_set>
+
+#include "dictionary_segment.hpp"
+#include "type_cast.hpp"
+
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+template <typename T>
+DictionarySegment<T>::DictionarySegment(const std::shared_ptr<AbstractSegment>& abstract_segment) {
+  for (auto index = ChunkOffset{0}; index < abstract_segment->size(); index++) {
+    const auto& value = type_cast<T>(abstract_segment->operator[](index));
```

Comment:
We should only use `[]` operator for debugging and tests. Using it in prod code yields to the fairly unclean function call `->operator[](index)`.

Reply:
Changed it, as in this case it's clear that we want to/can cast to ValueSegment pointer.
This may be interesting for a larger discussion though - if we aren't sure that there is not another segment that could be used as input to the DictionarySegment, and the ``operator[]`` is the only direct access implemented for all segments, then shouldn't we use that instead of making assumptions and casting?